File: ControlLib/src/main/java/org/marsroboticsassociation/controllib/motion/FunctionalCurveSegment.java
package org.marsroboticsassociation.controllib.motion;

import java.util.function.DoubleUnaryOperator;

/**
 * Analytic curve segment backed by value and slope functions over an absolute time interval.
 */
public final class FunctionalCurveSegment implements TrajectoryCurveSegment {

    private static final int EXTREMA_SAMPLES = 128;

    private final double startTime;
    private final double endTime;
    private final DoubleUnaryOperator valueFunction;
    private final DoubleUnaryOperator slopeFunction;

    public FunctionalCurveSegment(
            double startTime,
            double endTime,
            DoubleUnaryOperator valueFunction,
            DoubleUnaryOperator slopeFunction) {
        this.startTime = startTime;
        this.endTime = endTime;
        this.valueFunction = valueFunction;
        this.slopeFunction = slopeFunction;
    }

    @Override
    public double startTime() {
        return startTime;
    }

    @Override
    public double endTime() {
        return endTime;
    }

    @Override
    public double valueAt(double time) {
        return valueFunction.applyAsDouble(time);
    }

    @Override
    public double slopeAt(double time) {
        return slopeFunction.applyAsDouble(time);
    }

    @Override
    public double minValue() {
        return sampledExtremum(false);
    }

    @Override
    public double maxValue() {
        return sampledExtremum(true);
    }

    @Override
    public TrajectoryCurveSegment shiftedBy(double deltaTime) {
        return new FunctionalCurveSegment(
                startTime + deltaTime,
                endTime + deltaTime,
                time -> valueAt(time - deltaTime),
                time -> slopeAt(time - deltaTime));
    }

    @Override
    public TrajectoryCurveSegment clippedTo(double newStartTime, double newEndTime) {
        double clippedStart = Math.max(startTime, newStartTime);
        double clippedEnd = Math.min(endTime, newEndTime);
        if (clippedEnd <= clippedStart) return null;
        return new FunctionalCurveSegment(clippedStart, clippedEnd, this::valueAt, this::slopeAt);
    }

    private double sampledExtremum(boolean max) {
        if (duration() <= 0) return valueAt(startTime);
        double best = valueAt(startTime);
        double endValue = valueAt(endTime);
        best = max ? Math.max(best, endValue) : Math.min(best, endValue);
        for (int i = 1; i < EXTREMA_SAMPLES; i++) {
            double t = startTime + duration() * i / EXTREMA_SAMPLES;
            double value = valueAt(t);
            best = max ? Math.max(best, value) : Math.min(best, value);
        }
        return best;
    }
}

File: ControlLib/src/main/kotlin/org/marsroboticsassociation/controllib/motion/FunctionalCurveSegment.kt
package org.marsroboticsassociation.controllib.motion

import java.util.function.DoubleUnaryOperator
import kotlin.math.max
import kotlin.math.min

/** Analytic curve segment backed by value and slope functions over an absolute time interval. */
class FunctionalCurveSegment(
    private val startTime: Double,
    private val endTime: Double,
    private val valueFunction: DoubleUnaryOperator,
    private val slopeFunction: DoubleUnaryOperator,
) : TrajectoryCurveSegment {

    override fun startTime(): Double = startTime

    override fun endTime(): Double = endTime

    override fun valueAt(time: Double): Double = valueFunction.applyAsDouble(time)

    override fun slopeAt(time: Double): Double = slopeFunction.applyAsDouble(time)

    override fun minValue(): Double = sampledExtremum(false)

    override fun maxValue(): Double = sampledExtremum(true)

    override fun shiftedBy(deltaTime: Double): TrajectoryCurveSegment {
        return FunctionalCurveSegment(
            startTime + deltaTime,
            endTime + deltaTime,
            { time -> valueAt(time - deltaTime) },
            { time -> slopeAt(time - deltaTime) },
        )
    }

    override fun clippedTo(newStartTime: Double, newEndTime: Double): TrajectoryCurveSegment? {
        val clippedStart = max(startTime, newStartTime)
        val clippedEnd = min(endTime, newEndTime)
        if (clippedEnd <= clippedStart) return null
        return FunctionalCurveSegment(clippedStart, clippedEnd, this::valueAt, this::slopeAt)
    }

    private fun sampledExtremum(wantMax: Boolean): Double {
        if (duration() <= 0) return valueAt(startTime)
        var best = valueAt(startTime)
        val endValue = valueAt(endTime)
        best = if (wantMax) max(best, endValue) else min(best, endValue)
        for (i in 1 until EXTREMA_SAMPLES) {
            val t = startTime + duration() * i / EXTREMA_SAMPLES
            val value = valueAt(t)
            best = if (wantMax) max(best, value) else min(best, value)
        }
        return best
    }

    companion object {
        private const val EXTREMA_SAMPLES = 128
    }
}
